File: _submissions/19UCS159_Tharun/Day1/Q7.java
/*
Sort Binary Linked List : https://www.interviewbit.com/problems/sort-binary-linked-list/

Given a Linked List A consisting of N nodes.

The Linked List is binary i.e data values in the linked list nodes consist of only 0's and 1's.

You need to sort the linked list and return the new linked list.

NOTE:

Try to do it in constant space.

*/
// Definition for singly-linked list.
class ListNode {
  public int val;
  public ListNode next;

  ListNode(int x) {
    val = x;
    next = null;
  }
}

public class Q7 {
  public ListNode solve(ListNode A) {
    int zeroes = 0;
    ListNode curr = A;
    while (curr != null) {
      if (curr.val == 0)
        zeroes++;
      curr = curr.next;
    }
    curr = A;
    while (curr != null) {
      if (zeroes > 0) {
        curr.val = 0;
        zeroes--;
      } else {
        curr.val = 1;
      }
      curr = curr.next;
    }
    return A;
  }
}

File: _submissions/19UCS125_Rahul/Day1/q7.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
ListNode* Solution::solve(ListNode* A) {
    ListNode* head = A;
    ListNode *tail, *temp;
    temp = A;
    int n = 0;
    while(temp != NULL) {
        tail = temp;
        temp = temp->next;
        n++;
    }
    //cout<<n<<endl;
    temp = head;
    ListNode *prev = head;
    for(int i=0; i<n && temp->next != NULL; ++i) {
        if(temp->val == 1) {
            if(temp == head) {
                head = (temp->next != NULL) ? temp->next : head;
                prev = head;
                tail->next = temp;
                temp->next = NULL;
                tail = tail->next;
                temp = head;
            }
            else {
                prev->next = temp->next;
                temp->next = NULL;
                tail->next = temp;
                tail = tail->next;
                temp = prev->next;
            }
        }
        else {
            prev = temp;
            temp = temp->next;
        }
    }
    
    return head;
}
